test/e2e/apimachinery/garbage_collector.go

```diff
@@ -998,6 +998,109 @@ var _ = SIGDescribe("Garbage collector", func() {
 		}
 	})
 
+	It("should support orphan deletion of custom resources", func() {
+		config, err := framework.LoadConfig()
+		if err != nil {
+			framework.Failf("failed to load config: %v", err)
+		}
+
+		apiExtensionClient, err := apiextensionsclientset.NewForConfig(config)
+		if err != nil {
+			framework.Failf("failed to initialize apiExtensionClient: %v", err)
+		}
+
+		// Create a random custom resource definition and ensure it's available for
+		// use.
+		definition := apiextensionstestserver.NewRandomNameCustomResourceDefinition(apiextensionsv1beta1.ClusterScoped)
```

Comment:
I'm surprised this passes, I thought the GC only worked on namespace scoped resources?

Reply:
owner references don't have a namespace field, so they don't allow 
* cross-namespace references
* cluster-scoped -> namespaced references

They do allow:
* same namespace references
* namespaced -> cluster scoped references

Reply:
Owner reference was designed to not have namespace field, but there was use case / bug where dependent refers cluster-scoped owner. Fixed in: https://github.com/kubernetes/kubernetes/pull/57211

Reply:
Though it's not by design, I know GC at least supports cluster-scoped owner with namespaced dependents (see https://github.com/kubernetes/kubernetes/pull/57211). This test sets up cluster-scoped owner with cluster-scoped dependents, so it seems GC works in this case as well.

Maybe GC works for namespaced owner and cluster-scoped dependents in most cases. After all, GC's internal graph records the namespace, and indexed the graph nodes by UID.

Reply:
> Maybe GC works for namespaced owner and cluster-scoped dependents in most cases.

Specifying incorrect ownerref data and racing the uid index is definitely not safe, reliable, or supported. In cases where GC falls back to a live lookup of the owner before deleting, I think it would get a 404 from the incorrect (no namespace) reference, properly assume the pod with namespace "" and name "foo" did not exist, and delete the cluster scoped object.

Reply:
> In cases where GC falls back to a live lookup of the owner before deleting, I think it would get a 404 from the incorrect (no namespace) reference

Sure, in general it's not safe and not supported to have namespaced owner and cluster scoped dependents.

FWIW, the live lookup actually might work, because GC gets the namespace of the owner from the its internal graph, which is generated by the reflector in most cases, so the correct namespace will be there. I think the "virtual node" scenario is what is broken, where as liggitt described, the cluster-scoped child will be wrongly deleted.

Reply:
@lavalamp see https://github.com/kubernetes/kubernetes/issues/65200